pkg/expect/expect.go

```diff
@@ -187,9 +193,26 @@ func (ep *ExpectProcess) ExpectFunc(ctx context.Context, f func(string) bool) (s
 		}
 	}
 
+	select {
+	// NOTE: The ep.read() goroutine needs the mutex so we have to drain the
+	// log without lock.
```

Comment:
> so we have to drain the log without lock. 

Could you describe where we drain the log without lock? I thought ep.read() always drains the log with the mutex.

Reply:
Sorry for the unclear comment. 

https://github.com/etcd-io/etcd/blob/4364e7581b5f5fd664916fb6cb80e0a396b7e57d/pkg/expect/expect.go#L102-L138

If `ep.read()` just reads last line from pty, it will try to get lock and then append the log into `ep.lines`. So, if the `ep.Expect`  waits the `ep.readCloseCh` channel close with the lock, the `ep.read()` doesn't have a chance to get the lock and then the `ep.readCloseCh` won't be closed until `ep.Expect` releases the lock. It's deadlock.

Reply:
Thanks for the clarification! 

```
// NOTE: The ep.read() goroutine needs the mutex so we have to drain the
// log without lock.
```

I guess `ep.ExpectFunc` does not drain the log which causes my confusion. 

How about updating to?
```
// NOTE: we wait readCloseCh for ep.read() to complete draining the log before acquring the lock. 
```